realm/realm-library/src/main/java/io/realm/RealmBaseRecyclerAdapter.java

```diff
@@ -0,0 +1,195 @@
+            }
+        }
+    }
+
+    /**
+     * Clears all data from the adapter. This also deletes the objects from the underlying Realm.
+     */
+    public void clear() {
+        if (isDataValid()) {
+            getRealm().beginTransaction();
+            int dataSize = adapterData.size();
+            for (int i = dataSize - 1; i >= 0; i--) {
+                adapterData.deleteFromRealm(i);
+            }
+            getRealm().commitTransaction();
+
+            // There is no need to call notify methods if we have listener
+            if (!hasAutoUpdates) {
+                notifyDataSetChanged();
+            }
+        }
+    }
+
+    /**
+     * Returns data associated with this adapter.
+     *
+     * @return adapter data.
+     */
+    public OrderedRealmCollection<T> getData() {
+        return adapterData;
+    }
+
+    /**
+     * Updates the data associated to the Adapter. Useful when the query has been changed.
+     * If the query does not change you might consider using the automaticUpdate feature.
+     *
+     * @param data the new {@link OrderedRealmCollection} to display.
+     */
+    public void updateData(OrderedRealmCollection<T> data) {
+        if (hasAutoUpdates) {
+            if (adapterData != null) {
+                removeListener(adapterData);
+            }
+            if (data != null) {
+                addListener(data);
+            }
+        }
+
+        this.adapterData = data;
+        notifyDataSetChanged();
+    }
+
+    private boolean isDataValid() {
+        return adapterData != null && adapterData.isValid();
+    }
+
+    private BaseRealm getRealm() {
+        if (adapterData instanceof RealmResults) {
+            RealmResults realmResults = (RealmResults) adapterData;
+            return realmResults.realm;
+        } else if (adapterData instanceof RealmList) {
+            RealmList realmList = (RealmList) adapterData;
+            return realmList.realm;
+        } else {
+            throw new IllegalArgumentException("RealmCollection not supported: " + adapterData.getClass());
+        }
+    }
+
+    private void addListener(OrderedRealmCollection<T> data) {
+        if (data instanceof RealmResults) {
+            RealmResults realmResults = (RealmResults) data;
+            realmResults.realm.handlerController.addChangeListenerAsWeakReference(listener);
+        } else if (data instanceof RealmList) {
+            RealmList realmList = (RealmList) data;
+            realmList.realm.handlerController.addChangeListenerAsWeakReference(listener);
+        } else {
+            throw new IllegalArgumentException("RealmCollection not supported: " + data.getClass());
+        }
+    }
+
+    private void removeListener(OrderedRealmCollection<T> data) {
+        if (data instanceof RealmResults) {
+            RealmResults realmResults = (RealmResults) data;
+            realmResults.realm.handlerController.removeWeakChangeListener(listener);
```

Comment:
Unfortunately if I move` RealmBaseRecyclerAdapter` to `io.realm.android` it will not be possible to use `getRealm()` because `BaseRealm` class has package scope. Maybe for now the best solution is to use RealmResults inside adapter only without messy code like `realmResults.realm`. @cmelchior @beeender  What do you think?